File: apps/server/internal/modules/warls/session.go
package warls

import (
	"context"
	"encoding/json"
	"errors"

	deviceconsts "github.com/storganizer/server/internal/modules/devices/constants"

	"github.com/coder/websocket"
	"github.com/pocketbase/pocketbase/core"
)

// setMessage is the only JSON shape the session WebSocket accepts. The frontend
// publishes the full desired LED state for every device it wants to drive in
// one message; a device that drops out of `Frames` is cleared.
//
//	{"type":"set","frames":{"<deviceId>":[{"idx":3,"r":255,"g":0,"b":0}]}}
//	{"type":"set","frames":{}}                    // clear everything
type setMessage struct {
	Type   string             `json:"type"`
	Frames map[string][]frame `json:"frames"`
}

type frame struct {
	Idx int   `json:"idx"`
	R   uint8 `json:"r"`
	G   uint8 `json:"g"`
	B   uint8 `json:"b"`
}

// newSessionHandler returns a handler that upgrades the request to WebSocket
// and binds the connection to the given Registry. On disconnect, every device
// this session touched is cleared.
//
// Last-write-wins across sessions: a different session (or future HA
// integration) setting the same device URL replaces this session's frame.
func newSessionHandler(reg *Registry) func(*core.RequestEvent) error {
	return func(e *core.RequestEvent) error {
		conn, err := websocket.Accept(e.Response, e.Request, &websocket.AcceptOptions{
			// Browser dev servers run on a different origin than the
			// backend. We're inside the local network, so this is acceptable.
			InsecureSkipVerify: true,
		})
		if err != nil {
			return err
		}
		defer conn.Close(websocket.StatusInternalError, "closing")

		// Device URLs this session has currently set. Used to clear them all
		// on disconnect.
		ownedURLs := make(map[string]struct{})

		ctx := e.Request.Context()
		for {
			_, data, err := conn.Read(ctx)
			if err != nil {
				for url := range ownedURLs {
					reg.Clear(url)
				}
				if errors.Is(err, context.Canceled) ||
					websocket.CloseStatus(err) == websocket.StatusNormalClosure ||
					websocket.CloseStatus(err) == websocket.StatusGoingAway {
					return nil
				}
				return nil
			}

			var msg setMessage
			if err := json.Unmarshal(data, &msg); err != nil {
				continue
			}
			if msg.Type != "set" {
				continue
			}

			// Resolve device id → URL once per message. Build the new owned
			// set so we can clear any device that dropped out.
			nextOwned := make(map[string]struct{}, len(msg.Frames))
			for deviceID, frames := range msg.Frames {
				device, err := e.App.FindRecordById(deviceconsts.Collection, deviceID)
				if err != nil {
					continue
				}
				url := device.GetString(deviceconsts.FieldURL)
				if url == "" {
					continue
				}
				leds := make([]LED, 0, len(frames))
				for _, f := range frames {
					leds = append(leds, LED{Idx: f.Idx, R: f.R, G: f.G, B: f.B})
				}
				reg.Set(url, leds)
				nextOwned[url] = struct{}{}
			}
			for url := range ownedURLs {
				if _, kept := nextOwned[url]; !kept {
					reg.Clear(url)
				}
			}
			ownedURLs = nextOwned
		}
	}
}

File: apps/server/src/warls/session.py
"""WARLS WebSocket session.

The frontend publishes the full desired LED state for every device it wants to
drive in one message; a device that drops out of `frames` is cleared. On
disconnect, every device this session touched is turned off.

    {"type":"set","frames":{"<deviceId>":[{"idx":3,"r":255,"g":0,"b":0}]}}
    {"type":"set","frames":{}}     // clear everything
"""

from __future__ import annotations

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from src.core.database import SessionLocal
from src.devices.models import Device
from src.warls.registry import Registry
from src.wled.client import LEDColor

router = APIRouter()


async def _resolve_url(device_id: str) -> str | None:
    async with SessionLocal() as db:
        device = await db.get(Device, device_id)
        return device.url if device is not None else None


@router.websocket("/api/warls/stream")
async def warls_stream(websocket: WebSocket) -> None:
    registry: Registry = websocket.app.state.warls_registry
    await websocket.accept()

    owned_urls: set[str] = set()
    try:
        while True:
            msg = await websocket.receive_json()
            if msg.get("type") != "set":
                continue

            frames: dict = msg.get("frames") or {}
            next_owned: set[str] = set()
            for device_id, leds in frames.items():
                url = await _resolve_url(device_id)
                if not url:
                    continue
                colors = [
                    LEDColor(
                        index=int(f["idx"]),
                        r=int(f.get("r", 0)),
                        g=int(f.get("g", 0)),
                        b=int(f.get("b", 0)),
                    )
                    for f in leds
                ]
                registry.set(url, colors)
                next_owned.add(url)

            # Clear any device that dropped out of this frame.
            for url in owned_urls - next_owned:
                registry.clear(url)
            owned_urls = next_owned
    except WebSocketDisconnect:
        pass
    finally:
        for url in owned_urls:
            registry.clear(url)
